crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs

```diff
@@ -61,13 +66,33 @@ impl Rule for NoConstAssign {
         let model = ctx.model();
 
         let declared_binding = model.declaration(node)?;
-        if let Some(variable_declaration) = declared_binding
-            .syntax()
-            .ancestors()
-            .find_map(|ancestor| JsVariableDeclaration::cast_ref(&ancestor))
-        {
-            if variable_declaration.is_const() {
-                return Some(declared_binding.syntax().text_trimmed_range());
+
+        if let Some(possible_declarator) = declared_binding.syntax().ancestors().find(|node| {
+            !JsAnyObjectBindingPatternMember::can_cast(node.kind())
+                && !JsObjectBindingPatternPropertyList::can_cast(node.kind())
+                && !JsAnyArrayBindingPatternElement::can_cast(node.kind())
+                && !JsArrayBindingPatternElementList::can_cast(node.kind())
+                && !JsIdentifierBinding::can_cast(node.kind())
+        }) {
+            if JsVariableDeclarator::can_cast(possible_declarator.kind()) {
+                let possible_declaration = possible_declarator.parent()?;
+                if let Some(js_for_variable_declaration) =
+                    JsForVariableDeclaration::cast_ref(&possible_declaration)
+                {
+                    if js_for_variable_declaration.is_const() {
+                        return Some(declared_binding.syntax().text_trimmed_range());
+                    }
+                }
+
+                if let Some(js_variable_declaration) =
+                    JsVariableDeclaratorList::cast_ref(&possible_declaration)
+                        .and_then(|declaration| declaration.syntax().parent())
+                        .and_then(JsVariableDeclaration::cast)
+                {
+                    if js_variable_declaration.is_const() {
+                        return Some(declared_binding.syntax().text_trimmed_range());
+                    }
+                }
```

Comment:
Nit: You can use `else if` here because it is guaranteed that the `possible_declaration` isn't a `JsVariableDeclaratorList` if it is a `JsForVariableDeclaration`
```suggestion
                }

                else if let Some(js_variable_declaration) =
                    JsVariableDeclaratorList::cast_ref(&possible_declaration)
                        .and_then(|declaration| declaration.syntax().parent())
                        .and_then(JsVariableDeclaration::cast)
                {
                    if js_variable_declaration.is_const() {
                        return Some(declared_binding.syntax().text_trimmed_range());
                    }
                }
```